Google.Api.Generator/Generation/MethodDetails.cs

```diff
@@ -253,37 +253,112 @@ private static MethodDetails DetectPagination(ServiceDetails svc, MethodDescript
+            // - PageSizeCandidate should be named page_size
+            // - The repeated candidate should be the first in both orders
+            // - There should be 0 or more than 1 map candidates, to disambiguate with DiREGapic single-map case
+            //   OR The repeated candidate should have a field number of 1 
+            if (pageSizeCandidate.Name == "page_size" && repeatedCandidatesByDeclOrder[0] == repeatedCandidatesByNumOrder[0] && 
+                (repeatedCandidatesByNumOrder[0].FieldNumber == 1 || mapCandidates.Count != 1))
+            {
+                return new Paginated(svc, desc, repeatedCandidatesByDeclOrder[0], pageSizeCandidate.FieldNumber, pageTokenCandidate.FieldNumber);
+            }
+
+            // DiREGapic case where a return message has exactly one map
+            if (mapCandidates.Count == 1)
+            {
+                return new Paginated(svc, desc, mapCandidates.Single(), pageSizeCandidate.FieldNumber,
+                    pageTokenCandidate.FieldNumber);
+            }
+
+            // DiREGapic case where a return message has exactly one repeated field
+            // (pageSizeCandidate's name can be either "page_size" or "max_results")
+            if (repeatedCandidatesByDeclOrder.Count == 1 && !mapCandidates.Any())
+            {
+                return new Paginated(svc, desc, repeatedCandidatesByDeclOrder.Single(), pageSizeCandidate.FieldNumber, pageTokenCandidate.FieldNumber);
             }
-            return null;
 
-            bool IsCandidateField(FieldDescriptor field) =>
+            var errMsg = $"The method {desc.FullName} is selected as a pagination candidate " +
+                         $"but the configuration of the item response field candidates " +
+                         $"does not match any of the configurations we can generate.";
+                              
+            throw new InvalidOperationException(errMsg);
+
+            bool IsRepeatedCandidate(FieldDescriptor field) =>
                 field.IsRepeated && !field.IsMap && (field.FieldType == FieldType.Message || field.FieldType == FieldType.String);
         }
 
+        /// <summary>
+        /// Finds an appropriate candidate for a `page size`/`max results` request field.
+        /// That is a field that is named "page_size" or "max_results" and has a non-repeated Int32 type, if any.
+        /// If no such field exists, then simply a field that is named "page_size" or "max_results".
+        /// </summary>
+        private static FieldDescriptor FindPageSizeCandidate(MessageDescriptor input)
+        {
+            // Has the int32 page_size or int32 max_results field which defines the maximum number of paginated resources to return in the response
+            var pageSizeCandidate = input.FindFieldByName("page_size");
+            var maxResultsCandidate = input.FindFieldByName("max_results");
+
+            if (pageSizeCandidate is null && maxResultsCandidate is null)
```

Comment:
I'd be tempted to do all of this with a conditional operator:

```csharp
return pageSizeCandidate is null && maxResultsCandidate is null ? null
  : IsValidPageSizeCandidate(pageSizeCandidate) ? pageSizeCandidate
  : IsValidPageSizeCandidate(maxResultsCandidate) ? maxResultsCandidate
  : pageSizeCandidate ?? maxResultsCandidate;
```

But I'm more of a fan of the conditional operator than some other folks. No need to change, just something to consider.

Reply:
Also note that we don't actually need the first line at all - if they're both null, the rest of the conditions will end up returning null.

Reply:
👍